### DISCUSSION TITLE
How use ui.spinner in the event_handler?

### DISCUSSION THREAD

**Discussion by bobanovo** [zauberzeug/nicegui#816] [Q&A]
### Question

I would like to ask you for your help.
I have a simple page with splitter. On the left side I have table (ui.table), which contains data from API, and on the row I added slot on the q-tr for emitting own action and make a table.on on this handler. It works.
When I click on the row on the left, on the right side I want to show detailed data, which is based also on the external API. But, this API is a little slow. Therefore, I would like to show the spinner till the data will be loaded.

But, I don't use the spinner in the right way.

My first attempt was, to show ui.spinner in one row on the top of the right side , but, the spinner running still, and I cannot stop, in another else way, I set the variable spinner.visible=True for waiting for loading data, but the showing spinner waited to load data from API :-( and this it lost sense :-).

```py
def handle_action(msg) -> None:
        --> NOW I WANT START SPINNER
        print(msg)
        serverDetailData = getDetailOfServer_new (novaC, cinderC, msg['args'])
	labelServer.text=serverDetailData['serverdetail']['name']
		
	DetailPart.clear()
	with DetailPart:
		for rec in serverDetailData['serverdetail']:
			viewDict([{rec:serverDetailData['serverdetail'][rec]}],['status'])
		with DetailPart.add_slot('header'):
			ui.label('Server detail')
		DetailPart.value=True
        --> NOW I WANT HIDE SPINNER
```

What is the right way, how start the spinner, on the start handler_function, and stop at the end of handler_function?

**Comment by rodja** [comment]
Have a look at our [ffmpeg example](https://github.com/zauberzeug/nicegui/blob/main/examples/ffmpeg_extract_images/main.py). There we show a spinner while ffmpeg is creating the thumbnail images of uploaded videos in the background. And maybe FastAPI's great introduction and explanations about async/await are helpful: https://fastapi.tiangolo.com/async/

**Reply by bobanovo** [reply]
Thanks a lot, and sorry if this place is wrong place for my question/trouble... 
I tried do an example, see below, and in the console output is the correct behavior, but in the web ui is like synchronous behavior... spinner doesn't have time to show, and immediately is closed.

```python
import asyncio
import time
from nicegui import ui
 

def wait(seconds):
    start = time.time()
    while time.time() < start + seconds:
        pass

async def lib1():
    print("Start loading")
    with waiting:
        ui.spinner()
    ui.label("Start loading")
    res=[]
    for i in range (5):
        wait(1)
        res.append("-X-")
    waiting.clear()
    return res


async def click():
    print ('Start...')
    ui.label('Start...')
    
    res = await lib1()
    
    print (res)
    ui.label(res)
    print ('...Finish')
    ui.label ('...Finish')
    
ui.button('Start', on_click=click)
with ui.element('div') as waiting:
    pass

ui.run(port=5000)

```

What I do wrong?
Thx a lot!

**Reply by rodja** [reply]
You need to use `asyncio.sleep` and other awaitable coroutines.

**Comment by falkoschindler** [comment]
I think a very similar question has be answered over here: https://github.com/zauberzeug/nicegui/discussions/695.
